pkg/submitter/submitter.go

```diff
@@ -0,0 +1,573 @@
+		},
+			[]string{"slot"},
+		),
+		submitCost: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_cost",
+			Help:        "The current submit cost in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+		submitReward: promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_reward",
+			Help:        "The current reward in 1e18 eth",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		},
+			[]string{"slot"},
+		),
+	}
+
+	return submitter, submitter.resultCh, nil
+}
+
+func (s *Submitter) Start() error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			if s.resultCh != nil {
+				close(s.resultCh)
+			}
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			level.Info(s.logger).Log("msg", "submitter shutdown complete")
+			return s.ctx.Err()
+		case result := <-s.resultCh:
+			if s.lastSubmitCncl != nil {
+				s.lastSubmitCncl()
+			}
+			var ctx context.Context
+			ctx, s.lastSubmitCncl = context.WithCancel(s.ctx)
+
+			level.Info(s.logger).Log("msg", "received a solution",
+				"challenge", fmt.Sprintf("%x", result.Work.Challenge),
+				"solution", result.Nonce,
+				"difficulty", result.Work.Challenge.Difficulty,
+				"requestIDs", fmt.Sprintf("%+v", result.Work.Challenge.RequestIDs),
+			)
+			s.blockUntilTimeToSubmit(ctx)
+			s.handleSubmit(ctx, result)
+		}
+	}
+}
+
+func (s *Submitter) CancelPendingSubmit() {
+	if s.lastSubmitCncl != nil {
+		s.lastSubmitCncl()
+	}
+}
+
+func (s *Submitter) Stop() {
+	s.close()
+}
+
+func (s *Submitter) blockUntilTimeToSubmit(newChallengeReplace context.Context) {
+	var (
+		lastSubmit time.Duration
+		timestamp  *time.Time
+		err        error
+	)
+	for {
+		select {
+		case <-newChallengeReplace.Done():
+		default:
+		}
+		lastSubmit, timestamp, err = s.lastSubmit()
+		if err != nil {
+			level.Debug(s.logger).Log("msg", "checking last submit time", "err", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		break
+	}
+	if lastSubmit < s.cfg.Mine.MinSubmitPeriod.Duration {
+		level.Info(s.logger).Log("msg", "min transaction submit threshold hasn't passed", "minSubmitPeriod", s.cfg.Mine.MinSubmitPeriod, "lastSubmit", lastSubmit)
+		timeToSubmit, cncl := context.WithDeadline(newChallengeReplace, timestamp.Add(15*time.Minute))
+		defer cncl()
+		select {
+		case <-newChallengeReplace.Done():
+		case <-timeToSubmit.Done(): // 15min since last submit has passed to can unblock.
+		}
+	}
+}
+
+func (s *Submitter) handleSubmit(newChallengeReplace context.Context, result *mining.Result) {
+	go func(newChallengeReplace context.Context, result *mining.Result) {
+		ticker := time.NewTicker(1 * time.Second)
+		for {
+			select {
+			case <-newChallengeReplace.Done():
+				level.Info(s.logger).Log("msg", "canceled submit")
```

Comment:
please check that after every `newChallengeReplace.Done` we have one of these logs.
```suggestion
				level.Info(s.logger).Log("msg", "canceled pending submit")
```